Add ifs_as_logical_ops lint

barun511 opened the pull request:
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rust-clippy/pull/14904)*

Fixes https://github.com/rust-lang/rust-clippy/issues/14865 by adding the ifs_as_logical_ops lint.

Only adds the conjunctive version (transform if x else y => x && y) for now.

*Please write a short comment explaining your change (or "none" for internal only changes)*

changelog: [`ifs_as_logical_ops`] : "Warn about cases where `x && y` could be used in place of if conditions"

barun511 commented:
Hm, it doesn't look like my rebase really worked, let me have another go...

barun511 commented:
I'll take a look at merge conflicts once I've got a review, I think

barun511 commented on clippy_lints/src/ifs_as_logical_ops.rs:
```diff
@@ -0,0 +1,57 @@
+impl EarlyLintPass for IfsAsLogicalOps {
+    fn check_expr(&mut self, cx: &EarlyContext<'_>, e: &Expr) {
+        if let ExprKind::If(cond, cond_inner, Some(els)) = &e.kind
+            && let Some(inner_if_stmt) = cond_inner.stmts.first()
```
Just for my own understanding, by diverge you mean the case where we have an intentional panic of some sort here instead of a boolean, in which case ofc we shouldn't lint

barun511 commented:
Thanks for the review! I had a go at writing it with a late lint pass

Jarcho commented on clippy_lints/src/ifs_as_logical_ops.rs:
```diff
@@ -0,0 +1,67 @@
+            && if_block.stmts.len() == 0
+            && let Some(if_expr) = if_block.expr
+            // and does not diverge.
+            && is_never_expr(cx, if_expr).is_none()
```
The type check should also be after all the others as it will almost never catch anything and it's slightly slower.

barun511 commented:
Thanks! Responded

barun511 commented:
@rustbot ready

barun511 commented:
Addressed the comments, filled out the docs, and added some tests

samueltardieu reviewed (commented):
I think you have committed a `output.txt` file by mistake.

barun511 commented:
Thanks! Fixed

ada4a commented:
This suggestion (from Lintcheck) looks wrong:
```
warning: if expression that could be written as a logical and expression
  --> target/lintcheck/sources/bat-0.24.0/src/vscreen.rs:92:21
   |
92 | /                     if let Some((i, last)) = chars.last() {
93 | |                         // SAFETY: Always starts with ^[ and ends with m.
94 | |                         self.update_with_csi(last, &sequence[2..i])
95 | |                     } else {
96 | |                         false
97 | |                     }
   | |_____________________^ help: try: `let Some((i, last)) = chars.last() && self.update_with_csi(last, &sequence[2..i])`
```
You can't use `let <path> = <expr>` as a standalone bool (even though I often wish you could) – so you'll probably need to check that the first condition isn't an `if-let`/`while-let`

ada4a commented:
This ate some important stuff in the branch body:
```
warning: if expression that could be written as a logical and expression
   --> target/lintcheck/sources/clap_builder-4.5.9/src/parser/parser.rs:679:16
    |
679 |           } else if next.is_short() {
    |  ________________^
680 | |             // If this is a short flag, this is a new arg. But a single '-' by
681 | |             // itself is a value and typically means "stdin" on unix systems.
682 | |             debug!("Parser::is_new_arg: -<something> found");
...   |
687 | |             false
688 | |         }
    | |_________^ help: try: `next.is_short() && true`
    |
    = note: `--force-warn clippy::ifs-as-logical-ops` implied by `--force-warn clippy::nursery`
```

ada4a commented:
- the if-let problem, as described above
- the suggestion wants to change an `if x {} else if y {true} else {false}` into an `if x {} else y`, but that wouldn't work unless braces are added around the new `y`
```
warning: if expression that could be written as a logical and expression
   --> target/lintcheck/sources/encoding_rs-0.8.34/src/iso_2022_jp.rs:429:20
    |
429 |               } else if let Some(_) = jis0208_symbol_encode(bmp) {
    |  ____________________^
430 | |                 true
431 | |             } else {
432 | |                 false
433 | |             }
    | |_____________^ help: try: `let Some(_) = jis0208_symbol_encode(bmp) && true`
```

ada4a commented:
```
warning: if expression that could be written as a logical and expression
  --> target/lintcheck/sources/futures-util-0.3.31/src/stream/unfold.rs:84:9
   |
84 | /         if let UnfoldState::Empty = self.state {
85 | |             true
86 | |         } else {
87 | |             false
88 | |         }
   | |_________^ help: try: `let UnfoldState::Empty = self.state && true`
```

This overlaps with `clippy::match_like_matches_macro` – though I guess that won't be a problem once the if-let bug is fixed

ada4a commented:
> This overlaps with `clippy::match_like_matches_macro`

Similarly, I'm pretty sure the following pattern would get flagged as well:
```rs
if x { true } else { false}
```
which would overlap with [`clippy::needless_bool`](https://rust-lang.github.io/rust-clippy/master/index.html#needless_bool)

Jarcho commented:
> Curious why we need the empty expansions check at the end - can we not just check those on the way while checking our other conditions, as we're not going to lint anyway if we find something?

The order is just a perf thing. Getting the source text is a fair bit of work and the other conditions are more likely to filter things out.

> I also don't understand the significance of walking the span to the context, so the above question will probably resolve itself once I understand that.

Walking the context is needed so that all the spans are in the same context (roughly macro call). Relating the positions of two spans when they're in different contexts is completely bogus. Since they're different macro expansions there's no meaningful relation between the two.

> Yeah I did not consider the if x { if y { z } else { false } } else { false } case. In the linked example though, I imagine just removing the if would not work, because that would once again become "assign let directly to a boolean", but accept that need to solve this for the general case.
>
> As there are two expressions, the example does not trigger the internal lint for if y { z } else { false}, because that is also just y && z, so I wonder if we should directly suggest x && y && z for the case of if x { if y { z } else { false } } else { false } (though we would then have to do some recursive checks), or if it is acceptable to just suggest the separate lints.

Handling of the nested case is a 'nice to have', but it's not required. `--fix` will work correctly either way (it will fix one, rerun, and fix the other). As @ada4a already mentioned our tests don't handle this since we only run a single round, but that's not a problem for the lint in general.

ada4a commented on clippy_lints/src/ifs_as_logical_ops.rs:
```diff
@@ -0,0 +1,104 @@
+                return;
+            }
+            let mut sugg = Sugg::hir(cx, cond, "_");
+            let rhs_sugg = Sugg::hir(cx, if_expr, "_");
```
Yup!

ada4a commented on clippy_lints/src/ifs_as_logical_ops.rs:
```diff
@@ -0,0 +1,117 @@
+            let walked_option_if_block = walk_span_to_context(if_block.span, e.span.ctxt());
+            let walked_option_if_block_inner = walk_span_to_context(if_block_inner_expr.span, e.span.ctxt());
+            let walked_option_else_block = walk_span_to_context(else_block.span, e.span.ctxt());
+            let walked_option_else_block_inner = walk_span_to_context(else_block_inner_expr.span, e.span.ctxt());
```
1. `if_block` and `else_block` are checked to have the same context as `e`, so walking their spans should not be necessary
2. currently, the same check is present for `else_block_inner_expr`, but that's a bit weird -- why the asymmetry with `if_block_inner_expr`? I think it could make sense to allow both of them to come from a different context, and therefore walk their spans (I would like to hear maintainers' opinions on this)
3. for the spans that will end up being walked, the `ctxt` computed above could be reused (I _think_ repeatedly computing it is a bit expensive)
4. nit: imo it would make sense to inline these variables -- the names are kind of verbose, and they're only used once anyway

ada4a commented on clippy_lints/src/ifs_as_logical_ops.rs:
```diff
@@ -0,0 +1,117 @@
+            && let ExprKind::Block(if_block, _label) = cond_inner.kind
+            // Make sure the if block is not an if-let block.
+            && let ExprKind::DropTemps(_) = if_cond.kind
+            // Check if the if-block has only a return statement
```
```suggestion
            // Check if the if-block has only a trailing expression
```

ada4a reviewed (commented):
Looking pretty good now:) Left some smaller comments

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/14904/60a978d2e9c7b49f21de9d665365427882e781ae..1fb222979988893615854c40130c5b3ede5cb72f)*

barun511 commented on tests/ui/ifs_as_logical_ops.rs:
```diff
@@ -0,0 +1,95 @@
+#![warn(clippy::ifs_as_logical_ops)]
+#![allow(clippy::needless_bool)]
+#![allow(clippy::match_like_matches_macro)]
```
Fixed properly now

barun511 commented on tests/ui/ifs_as_logical_ops.rs:
```diff
@@ -0,0 +1,134 @@
+}
+
+fn example_with_if_let_does_not_lint(b1: bool, b2: Option<i32>) -> bool {
+    if let Some(30) = b2 { true } else { false }
```
Yeah, this still won't lint, but at least the test is more useful now (and now we no longer need to expect match like matches!)

barun511 commented on tests/ui/ifs_as_logical_ops.rs:
```diff
@@ -0,0 +1,134 @@
+}
+
+fn example_with_if_let_does_not_lint(b1: bool, b2: Option<i32>) -> bool {
+    if let Some(30) = b2 { true } else { false }
```
Neither of these get linted currently, added tests for both!

ada4a commented on clippy_lints/src/ifs_as_logical_ops.rs:
```diff
@@ -0,0 +1,117 @@
+            && ctxt == else_block.span.ctxt()
+            && ctxt == else_block_inner_expr.span.ctxt()
+            && ctxt == lit.span.ctxt()
+            && !ctxt.in_external_macro(cx.tcx.sess().source_map())
```
yeah having two traits for that is a bit confusing...

ada4a commented on output.txt:
```diff
@@ -0,0 +1,4 @@
+    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.01s
```
this sneaked in again^^

ada4a reviewed (commented):
I think I don't have anything more to add, looks pretty much perfect:)

It would be great if you could rebase on the latest master so that Lintcheck could run -- maybe we've missed some FPs after all

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/14904/60a978d2e9c7b49f21de9d665365427882e781ae..970a41b92522f7f8788918995335a747f46a3b3a)*

barun511 commented:
Yep that works totally fine.

ada4a commented on clippy_lints/src/ifs_as_logical_ops.rs:
```diff
@@ -0,0 +1,109 @@
+        if let ExprKind::If(if_cond, cond_inner, Some(els)) = e.kind
+            && let ExprKind::Block(if_block, _label) = cond_inner.kind
+            // Make sure the if block is not an if-let block.
+            && let Some(_) = higher::If::hir(e)
```
`If::hir` returns you `cond`/`then`/else`, so you might as well use them (instead of the first line of the lint)

barun511 replied:
Good point, thank you!

ada4a commented on clippy_lints/src/ifs_as_logical_ops.rs:
```diff
@@ -0,0 +1,110 @@
+impl<'tcx> LateLintPass<'tcx> for IfsAsLogicalOps {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, e: &'tcx Expr<'tcx>) {
+        // Make sure the if block is not an if-let block.
+        if let Some(If {cond: if_cond, then: cond_inner, r#else}) = If::hir(e)
```
a very nitty nit: you could destructure `else` here directly:
```suggestion
        if let Some(If {cond: if_cond, then: cond_inner, r#else: Some(els)}) = If::hir(e)
```
which avoids the need to do that separately below.

Only pointing this out because you used to do that when you were using `ExprKind::If`

ada4a commented on clippy_lints/src/ifs_as_logical_ops.rs:
```diff
@@ -0,0 +1,109 @@
+            && let ExprKind::Block(if_block, _label) = cond_inner.kind
+            // Check if the if-block has only a trailing expression
+            && if_block.stmts.is_empty()
+            && let Some(if_block_inner_expr) = if_block.expr
```
The first block of an if is often called `then`, I think that's a pretty good convention
```suggestion
        if let Some(If {cond: if_cond, then, r#else: Some(els)}) = If::hir(e)
            && let ExprKind::Block(then_block, _label) = then.kind
            // Check if the then-block has only a trailing expression
            && then_block.stmts.is_empty()
            && let Some(then_block_inner_expr) = then_block.expr
```

ada4a reviewed (commented):
Everything looks pretty good by now, I don't think I have much more to add^^

But I assume what you actually wanted was a final approval. Unfortunately I don't have the right to give those (nor would I feel confident to do so), so you might need to wait a bit for an actual reviewer^^

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/14904/ad7fc4f4684d825ad492f5b0f487fa601f7ac272..bb9f04ff415afe24ac216e65a2d938cfc6fb902e)*

barun511 commented:
@rustbot help

barun511 commented:
@rustbot r?

ada4a commented:
I got a notification about a comment from @zayutaha, but it seems to no longer be there? Anyway, I think this should be pretty much ready, so @llogiq I'd say we can go on and start the FCP?

llogiq reviewed (commented):
This looks good modulo the caveats @ada4a and @jarcho already mentioned. I think this could be a `pedantic` lint unless there are any problems.

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/14904/ad7fc4f4684d825ad492f5b0f487fa601f7ac272..bb9f04ff415afe24ac216e65a2d938cfc6fb902e)*

barun511 commented on tests/ui/ifs_as_logical_ops.rs:
```diff
@@ -0,0 +1,147 @@
+}
+
+fn basic_lint_test(b1: bool, b2: bool) -> bool {
+    if b1 { b2 } else { false }
```
Yes! Though of course that does not lint as we should not simplify away the cfg!

llogiq replied:
Exactly. If the lint would trigger, that would constitute a false positive.

barun511 commented:
1. Fixed @ada4a's comment in the unfixable file to extend the lint test declaration
2. Fixed @ada4a's comment about renaming if_block -> then_block
3. Fixed @llogiq's comment about linting the lint decription, and changing the type of lint
4. Fixed @llogiq's observation that we only check comments before the expression and not after it (actually true of both if and else conditions)
5. Weird thing, added the proc macro tests, but they don't lint even without my is_ext_proc_macro note, so the test just passes anyway. Is there a better way to get the test to fail? I followed instructions [here](https://doc.rust-lang.org/clippy/development/macro_expansions.html#the-is_from_proc_macro-function), and also looked at other tests, and they seemed to do similiar things 

I reorganized all my commits so you should only need to look at the latest one to see my changes, and all the previous mess of commits I've squashed into one

ada4a reviewed (changes requested):
It looks like `.DS_Store` slipped through. You might want to add it to your global gitignore to avoid this problem in the future

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/14904/e492d022858076a414ad42139e05cb43e0d238b6..009e0c6d4f5074b8531da52fdda7226faef9e640)*

ada4a commented on clippy_lints/src/ifs_as_logical_ops.rs:
```diff
@@ -33,51 +34,51 @@ declare_clippy_lint! {
             && ctxt == else_block_inner_expr.span.ctxt()
             && ctxt == lit.span.ctxt()
             && !ctxt.in_external_macro(cx.sess().source_map())
+        // && !is_from_proc_macro(cx,e)
```
not sure why this is, sorry..

samueltardieu commented on clippy_lints/src/ifs_as_logical_ops.rs:
````diff
@@ -0,0 +1,115 @@
+    ///     b1 && a()
+    /// }
+    /// ```
+    #[clippy::version = "1.89.0"]
````
Not even that, it should be "1.96.0".
